enhancements/cluster-logging/elasticsearch-cert-management.md

```diff
@@ -0,0 +1,158 @@
+
+### User Stories
+
+#### As an user of EO, I do not want to have to generate and provide secrets for the Elasticsearch cluster
+
+We want to further simplify using the Elasticsearch Operator and create a single source of truth for the certs the cluster uses.
+
+#### As a consumer of EO, I want a mechanism to have certificates injected into a secret for me to continue mTLS communcation with my Elasticsearch cluster
+
+We want to continue allowing mTLS as this supports the Cluster Logging and Jaeger use cases, however they shouldn't need to do more than notify they need certs injected by EO.
+
+### Implementation Details
+
+#### Assumptions
+
+* The Elasticsearch Operator will implement this in golang (rather than the current script used in CLO)
+
+* The operator will need to be able to store generated certificates for each namespace/unique cluster so that there is no security concerns of clusters cross communicating.
+
+#### Security
+
+Ensuring that signing certificates and configurations are isolated for different clusters is a must for this implementation. It will also need to be able to do this thread safe manner.
+
+### Risks and Mitigations
+
+## Design Details
+
+From a user perspective, there should be no change in any of the CRDs. The only major changes would be no longer generating secrets for EO to use, and needing to annotate a secret (user provided) in the case they want to communicate with ES in a mTLS way.
+
+### secret annotation
+
+We will add an annotation value to check for on secrets in namespaces where an es CR exists:
+
+```
+apiVersion: v1
+kind: Secret
+metadata:
+  name: fluentd
+  annotations:
+    secret.openshift.io/elasticsearch-cert-component-name: system.logging.fluentd
+type: Opaque
+data: {}
+```
+
+Where the value will be optional (defaults to the name of the secret) and corresponds to the CN used for the generated certificate.
+
+### Test Plan
+
+#### Unit Testing
+
+* We will need to ensure we have unit tests that verify certificates are generated by the EO
+
+* We will need to ensure we have unit tests that verify secrets are generated by EO with the correct certificates
+
+* We will need to ensure we have unit tests that verify EO generates and injects for annotated secrets
+
+#### Integration and E2E tests
+
+* Given we are moving logic from one operator to another, we need to ensure that there are no regressions with communication between Fluentd, Kibana, IM cronjobs, and Elasticsearch (this should be handled by our smoke test).
+
+### Graduation Criteria
+
+#### GA
+
+* End user documentation
+* Sufficient test coverage
+* Sufficient LTS upgrade strategy
+
+### Version Skew Strategy
+
+We need to have a proper upgrade strategy from our LTS version to the version this is released on (currently 4.6).
+
+## Implementation History
+
+| release|Description|
+|---|---|
+|5.2| **GA** - Initial release
+
+## Drawbacks
+
+None
```

Comment:
Arguably this EP doesn't go far enough, as the proposed functionality is not specific to EO, cluster logging or Jaeger, but belongs rather in in a OCP-cluster-wide CA.

Reply:
can you expand more on this? we've tried to get mTLS functionality into the OCP cert signing service and it seems that isn't going to be implemented so we would need to still sign our own certs

Reply:
I don't really know anything about the OCP cert signing service, it's just an observation that what we're talking about in this EP is a piece of generic functionality and there may be other components that sign their own certs out there that could use this. But if you've been down that road and its a dead end (for now?) so be it. It might be worth mentioning this in the EP.

Reply:
@syedriko does what i added address your comments?

Reply:
@ewolinetz In the Drawbacks section, yes ;) I think we really need to nail down the contents of the secrets ^^^ EO is going to populate as it's part of the contract with the its consumers.

Reply:
@syedriko I'm not sure I follow... it would be the same as they are currently populated today in CLO because the purpose is that of communicating with ES with mTLS (and ES will expect certain domain name patterns so that they are correctly matched to backing roles for authn/authz)

Reply:
@ewolinetz The contents of the secrets CLO populates is irregular - the names and number of CAs, the names of the cert and key are different in different secrets. I captured these at https://github.com/openshift/cluster-logging-operator/blob/c36a5589e7bc272f44153886ab1e4aba645d0638/pkg/k8shandler/certs.go#L73. I'm figuring EO will populate secrets in a uniform way, in which case we need to document how.

Reply:
@syedriko ack, i updated the document to describe the field names we'll use in the secrets for anyone that is opting in to have client certs generated for them. I believe it matches what is expected for the log forward use case.

Also, given I've updated the intention to make using this an opt-in configuration (by annotation on the elasticsearch CR) we don't have to worry about breaking anyone prior to them being ready to make all necessary changes that would go along with using certs managed by EO.